clippy_lints/src/raw_strings.rs

```diff
@@ -0,0 +1,123 @@
+    /// idiomatic than a string literal, so it's opt-in.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let r = r"Hello, world!";
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let r = "Hello, world!";
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub NEEDLESS_RAW_STRING,
+    restriction,
+    "suggests using a string literal when a raw string literal is unnecessary"
+}
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for raw string literals with an unnecessary amount of hashes around them.
+    ///
+    /// ### Why is this bad?
+    /// It's just unnecessary, and makes it look like there's more escaping needed than is actually
+    /// necessary.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let r = r###"Hello, "world"!"###;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let r = r#"Hello, "world"!"#;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub NEEDLESS_RAW_STRING_HASHES,
+    style,
+    "suggests reducing the number of hashes around a raw string literal"
+}
+impl_lint_pass!(RawStrings => [NEEDLESS_RAW_STRING, NEEDLESS_RAW_STRING_HASHES]);
+
+pub struct RawStrings {
+    pub needless_raw_string_hashes_allow_one: bool,
+}
+
+impl EarlyLintPass for RawStrings {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if !in_external_macro(cx.sess(), expr.span)
+            && let ExprKind::Lit(lit) = expr.kind
+            && let LitKind::StrRaw(num) | LitKind::ByteStrRaw(num) | LitKind::CStrRaw(num) = lit.kind
+        {
+            let prefix = match lit.kind {
+                LitKind::StrRaw(..) => "r",
+                LitKind::ByteStrRaw(..) => "br",
+                LitKind::CStrRaw(..) => "cr",
+                _ => unreachable!(),
+            };
+            if !snippet(cx, expr.span, prefix).trim().starts_with(prefix) {
+                return;
+            }
+
+            if !lit.symbol.as_str().contains(['\\', '"']) {
+                span_lint_and_sugg(
+                    cx,
+                    NEEDLESS_RAW_STRING,
+                    expr.span,
+                    "unnecessary raw string literal",
+                    "try",
+                    format!("{}\"{}\"", prefix.replace('r', ""), lit.symbol),
+                    Applicability::MachineApplicable,
+                );
+
+                return;
+            }
+
+            #[expect(clippy::cast_possible_truncation)]
+            let req = lit
+                .symbol
+                .as_str()
+                .as_bytes()
+                .split(|&b| b == b'"')
+                .skip(1)
+                .map_while(|bs| {
+                    let max = num;
+                    let num = bs.iter().take_while(|&&b| b == b'#').count() as u8;
+                    if num != max {
+                        return Some(num + 1);
+                    }
+
+                    None
+                })
```

Comment:
I don't think this fixes the issue. We know from the lexer that the number of hashes inside the literal will always be strictly smaller than the number of hashes surrounding the literal, so the `num == max` case can never occur. This code will do the same thing the previous one did.

In general, any `while`-based iterators don't work for this task, since they will always allow all items that match some condition. What we want to do is match the _first_ occurrence of `hash_length = num-1`, but not any of the following such occurrences.

Also keep in mind short circuiting the `split` is the same kind of issue. Ideally, searching the literal should stop as soon as the last necessary `#` is found, not when the following `"` is encountered.

Reply:
> This code will do the same thing the previous one did.

This was my bad actually, it was meant to check `num - 1`.

> In general, any while-based iterators don't work for this task,

I don't think this is true (with `num - 1`). `map_while` returns the value from the predicate, once `None` is returned, it will stop searching because it returns `None` from `next`, which would be as if it ran out of items (thus, it will stop iterating). You can see this in its implementation: https://doc.rust-lang.org/src/core/iter/adapters/map_while.rs.html#40-44. Since `None` is only returned in the case that it equals `max - 1`, the last element will either have any number of hashes (if it finished iterating) or `max - 1` (if it returned `None`).

Though it should be possible to remove `max`, by checking whether the last element is `max - 1` before running `max` on the rest.

Reply:
I just benchmarked it; for long strings (ran 1,000,000 times, as well as a couple dozen times to remove any potential randomness) it's *way* slower, UNLESS there's a `"` with `max - 1` hashes, in which case it's between 10000%-10% the speed of the original implementation (depending on where it is in the string). The performance hit for where it's near the end is probably due to it needing to call a closure, over and over again.

The difference is not large enough for other strings (but not small enough to be unnoticeable) that it's worth it imo in its current implementation.

Reply:
> This was my bad actually, it was meant to check `num - 1`.

You've swapped the `num` and `max` variables in the recent push, so it should be `max - 1` :P the code currently does the same thing again.

But, changing it to map take while `num != max - 1`, what I think you intended, would not work either. In that case, the iterator would finish without returning `num`, so it would not get considered in the max. This is what I meant about matching the same conditions: with `while` can either get all `max-1` and no `max`, or we can get all of the `max`. But we want to find just one of the `max` and stop there. It could be that I'm missing something though, let me know.

> The difference is not large enough for other strings (but not small enough to be unnoticeable) that it's worth it imo in its current implementation.

I found the difference to be unnoticeable with the new suggestion; perhaps you can try it with your benchmarks. If it doesn't look good, maybe my original line is fine to keep.

Reply:
> You've swapped the num and max variables in the recent push

I should not code while half asleep.

> But we want to find just one of the max and stop there.

This is what it does; the `max` would stop on the final element, which happens to be `max - 1`.

Reply:
That isn't true. It returned `num + 1` before.

Reply:
I see what you mean.